state/immutable-trie/storage.go

```diff
@@ -17,81 +16,39 @@ var (
 	codePrefix = []byte("code")
 )
 
-type Batch interface {
-	Put(k, v []byte)
-	Write()
-}
-
 // Storage stores the trie
 type Storage interface {
-	Put(k, v []byte)
-	Get(k []byte) ([]byte, bool)
-	Batch() Batch
+	kvdb.KVBatchStorage
+
 	SetCode(hash types.Hash, code []byte)
 	GetCode(hash types.Hash) ([]byte, bool)
-
-	Close() error
 }
 
 // KVStorage is a k/v storage on memory using leveldb
-type KVStorage struct {
-	db *leveldb.DB
-}
-
-// KVBatch is a batch write for leveldb
-type KVBatch struct {
-	db    *leveldb.DB
-	batch *leveldb.Batch
-}
-
-func (b *KVBatch) Put(k, v []byte) {
-	b.batch.Put(k, v)
+type kvStorage struct {
+	kvdb.KVBatchStorage
```

Comment:
Don't ever nest interface to struct, it was a bad practice. Especially use low module interface.

You could verify interface compliance using [uber-go golang guide](https://github.com/uber-go/guide/blob/master/style.md#verify-interface-compliance), which compiler could do those verification job for you.